A87-mtls-spiffe-support.md

```diff
@@ -0,0 +1,260 @@
+     FileWatcherCertificateProviderOptions& set_root_cert_path(absl::string_view root_cert_path);
+     FileWatcherCertificateProviderOptions& set_spiffe_bundle_map_path(absl::string_view spiffe_bundle_map_path);
+     FileWatcherCertificateProviderOptions& set_private_key_path(absl::string_view private_key_path);
+
+
+   private:
+     std::string private_key_path_;
+     std::string identity_certificate_path_;
+     std::string root_cert_path_;
+     std::string spiffe_bundle_map_path_;
+     unsigned int refresh_interval_sec_;
+}
+
+class FileWatcherCertificateProvider final .... {
+   // existing class definitions
+   FileWatcherCertificateProvider(const FileWatcherCertificateProviderOptions& options);
+}
+```
+
+#### Java
+
+In Java, we need new API to work with [SPIFFE] and [SPIFFE bundle
+format].  A new `SpiffeUtil` will be developed:
+
+```java
+class SpiffeUtil {
+   Optional<SpiffeId> extractSpiffeId(X509Certificate[] certChain);
+   SpiffeBundle loadTrustBundleFromFile(String trustBundleFile);
+}
+class SpiffeId {
+   String getTrustDomain();
+   String getPath();
+}
+class SpiffeBundle {
+   Map<String, List<X509Certificates>> getBundleMap();
+}
+```
+
+For the xDS functionality described above, the `XdsX509TrustManager`
+will gain the ability to be instantiated with a SPIFFE trust bundle
+map.  In this case, it will use the SPIFFE trust bundle certificates as
+trusted roots.  If the `XdsX509TrustManager` is instantiated using CA
+certificates (existing functionality), then it'll continue to use them
+exactly as it does today.  The new APIs will look like this:
+
+```java
+class XdsTrustManagerFactory {
+   XdsTrustManagerFactory(Map<String, List<X509Certificates>>);
+   XdsTrustManagerFactory(X509Certificates[]);
+}
+class XdsX509TrustManager {
+   XdsX509TrustManager(Map<String, XdsX509ExtendedTrustManager>);
+   XdsTrustManagerFactory(XdsX509ExtendedTrustManager);
+}
+```
+
+We'll also adjust existing hierarchies of `Watcher` and
+`CertificateProvider` to support the new SPIFFE trust bundle map.
+
+For the AdvancedTls case in Java, the API changes will look as follows:
+In AdvancedTlsX509TrustManager.java, the API must allow configuring and reloading a SPIFFE Bundle Map.
+There are currently three functions that configure the root trust certificate
+```
+public void updateTrustCredentials(X509Certificate[] trustCerts) throws IOException,
+   GeneralSecurityException
+
+public void updateTrustCredentials(File trustCertFile) throws IOException,
+   GeneralSecurityException
+
+public Closeable updateTrustCredentials(File trustCertFile, long period, TimeUnit unit,
+   ScheduledExecutorService executor) throws IOException, GeneralSecurityException
+```
+
+We will add three similar new functions for updating the SPIIFE root of trust
+```
+public void updateSpiffeTrustBundle(SpiffeBundle spiffeBundle) throws IOException,
```

Comment:
`SpiffeBundle` is internal API, so we shouldn't expose it here. Unless you were thinking we would make it public, in which case, I'd probably still argue against it since then we'd be signing up to support a full Spiffe parsing API.

I think instead we can just drop this method and not have it at all, or have an InputStream based version.

Reply:
Sounds good.